### PR TITLE
Fix WAL log data corruption #8723

### PR DESCRIPTION
**Pull request by eharry**
Fix WAL log data corruption when using DBOptions.manual_wal_flush(true) and WriteOptions.sync(true) together (#8723)

### CODE REVIEW

File: file/writable_file_writer.cc
Comment (riversand963) [inline file/writable_file_writer.cc:474]: I think the internal linter will recommend using `std::array` instead of raw array, let's just make the linter happy given the extra effort is trivial.

File: db/db_impl/db_impl_write.cc
Comment (riversand963) [inline db/db_impl/db_impl_write.cc:1130]: Nit: The FlushWAL

File: db/db_impl/db_impl_write.cc
Comment (riversand963) [inline db/db_impl/db_impl_write.cc:1132]: Nit: the log file may get data corruption

Comment (riversand963) [reply db/db_impl/db_impl_write.cc:1132]: Is 'duplicate record' a confirmed case? If the behavior is undefined, I think we can just say "data corruption".

File: db/db_test.cc
Comment (riversand963) [inline db/db_test.cc:4068]: It's not reliable to use `sleep` to cause certain interleaving between concurrent threads.
I may need to read the code more and understand this test better before giving a better suggestion that will use TEST_SYNC_POINTS.

**Review by riversand963 — COMMENTED** [review]
Thanks @eharry for the fix. I left a few comments.
btw, I think this fix is worth a note in the HISTORY.md, and the description could include a little more detail.

File: db/db_impl/db_impl_write.cc
Comment (eharry) [reply db/db_impl/db_impl_write.cc:1132]: > Is 'duplicate record' a confirmed case? If the behavior is undefined, I think we can just say "data corruption".

yes, the duplicate record is confirmed case. In the db_test.cc, when (splitWriteAndSlowSecondWrite: false, manual_wal_flush: true, sync: true), the test always failed by duplicate record.

File: db/db_test.cc
Comment (riversand963) [inline db/db_test.cc:4057]: A general suggestion before diving into the detailed interleaving of the threads:
since this test covers a combination of configurations, it will be worthwhile to define a new parameterized GTest fixture e.g. `DBWALFlushTest` (or something similar) so that we can improve readability.

File: db/db_test.cc
Comment (riversand963) [inline db/db_test.cc:4087]: We should assert on the result of `Put`.

File: db/db_test.cc
Comment (riversand963) [inline db/db_test.cc:4097]: We should assert on the result of `WriteImpl`

File: db/db_test.cc
Comment (riversand963) [inline db/db_test.cc:4104]: We should assert on the result of `FlushWAL`

File: db/db_test.cc
Comment (riversand963) [inline db/db_test.cc:4116]: We can simply do `ASSERT_EQ("b" + istr, Get("a" + istr))`

File: file/writable_file_writer.cc
Comment (riversand963) [inline file/writable_file_writer.cc:476]: Should we need a test sync point, it's recommended to name it `WritableFileWriter::WriteBuffered:AfterComputingAllowedSize`.

**Comment by riversand963** [comment]
Sorry about this late comment, but I am wondering whether we can simplify the test for this fix.

I have a simpler repro for the race condition: https://gist.github.com/riversand963/741e1ea092f6a32bdc3ec3b04dd3f6ee. I verified that TSAN can catch the bug with high probability.

The original bug is undefined behavior caused by unprotected concurrent access to some common data structures, so I guess we can try our best to cause the bug, but there does not seem a guarantee.

File: file/writable_file_writer.cc
Comment (eharry) [reply file/writable_file_writer.cc]: I just remove this call_back_para. 
There is a simple way to reproduce a WAL corruption by doing concurrent write and FlushWAL.  It is not necessary to reproduce another WAL corruption issue with a complicated  code.

File: db/db_impl/db_impl_write.cc
Comment (eharry) [reply db/db_impl/db_impl_write.cc]: already fix it

File: db/db_test.cc
Comment (eharry) [reply db/db_test.cc]: I just remove this call_back_para.
There is a simple way to reproduce a WAL corruption by doing concurrent write and FlushWAL. It is not necessary to reproduce another WAL corruption issue with a complicated code.

File: db/db_test.cc
Comment (eharry) [reply db/db_test.cc]: fix it,  add new test ManualFlushWalAndWriteRace

File: db/db_test.cc
Comment (eharry) [reply db/db_test.cc]: That diff is code format change.  Now,   add a new ManualFlushWalAndWriteRace test. Current diff does not have this code format change.

File: db/db_test.cc
Comment (eharry) [reply db/db_test.cc]: That diff is code format change.  Now,   add a new ManualFlushWalAndWriteRace test. Current diff does not have this code format change.

File: db/db_test.cc
Comment (eharry) [reply db/db_test.cc]: That diff is code format change.  Now,   add a new ManualFlushWalAndWriteRace test. Current diff does not have this code format change.

File: db/db_test.cc
Comment (eharry) [reply db/db_test.cc]: That diff is code format change.  Now,   add a new ManualFlushWalAndWriteRace test. Current diff does not have this code format change.

File: db/db_test.cc
Comment (eharry) [reply db/db_test.cc]: fix it, remove the split write operation, only use concurrent write to reproduce this issue.

File: file/writable_file_writer.cc
Comment (eharry) [reply file/writable_file_writer.cc]: fix it, remove the split write operation, only use concurrent write to reproduce this issue.

File: db/db_impl/db_impl_write.cc
Comment (riversand963) [inline db/db_impl/db_impl_write.cc:1128]: Thanks for the comment here. Just realized that the comment in line 1120 and so on will need to be updated due to this PR. Could you update it?

File: db/db_test.cc
Comment (riversand963) [inline db/db_test.cc:4102]: iiuc, this test failure will be caught with a probability. Maybe worth a comment here.

File: db/db_test.cc
Comment (riversand963) [inline db/db_test.cc:4135]: `ropt` is default-constructed. Here we can simply do `ASSERT_EQ("value1", Get("foo1"))` in which `Get()` is a wrapper of `DB::Get`.

**Review by riversand963 — COMMENTED** [review]
Thanks @eharry for the fix. Mostly LGTM except a few minor comments.

File: db/db_test.cc
Comment (eharry) [reply db/db_test.cc:4102]: Thanks for review. Already add a comments.

File: db/db_test.cc
Comment (eharry) [reply db/db_test.cc]: Thanks for review. Already use the Get().

File: db/db_impl/db_impl_write.cc
Comment (eharry) [reply db/db_impl/db_impl_write.cc:1128]: Thanks for review.
Line 1120 describes why it is not necessary to add _mutex when accessing _logs. That description is appropriate. 
This root cause of this issue is file->Sync running without locked log_write_mutex_, not accessing _logs.
So I did not modify the comment on line 1120

**Review by riversand963 — APPROVED** [review]

File: db/db_impl/db_impl_write.cc
Comment (eharry) [reply db/db_impl/db_impl_write.cc]: Thanks for review.  Already fix it.

Comment (riversand963) [reply db/db_impl/db_impl_write.cc]: Thanks for the fix, will merge early this week. In the meantime, it would be great if you can rebase to reduce the chance of conflicts with others' code changes. Once we get an approval on Phabricator, we can merge it.

**Pull request closed without merging**
